core/src/plugins/filed/python/postgresql/bareos-fd-postgresql.py

```diff
@@ -0,0 +1,1043 @@
+# 02110-1301, USA.
+#
+# Author: Bareos Team
+#
+"""bareos-fd-postgresql is a plugin to backup PostgreSQL cluster
+with bareos-fd-postgresql
+"""
+import os
+import io
+import json
+import getpass
+from sys import version_info
+import time
+import datetime
+import dateutil
+import dateutil.parser
+import bareosfd
+from bareosfd import *
+
+from BareosFdPluginLocalFilesBaseclass import BareosFdPluginLocalFilesBaseclass
+
+# Provided by the Bareos FD Python plugin interface
+
+# This module contains the wrapper functions called by the Bareos-FD, the
+# functions call the corresponding methods from your plugin class
+
+from BareosFdWrapper import *  # noqa
+
+
+try:
+    import pg8000
+
+    # check if pg8000 module is new enough
+    major, minor, patch = pg8000.__version__.split(".")
+    if int(major) < 1 or (int(major) == 1 and int(minor) < 16):
+        raise ValueError(f"pg8000 module version ({pg8000.__version__}) is lower than required version 1.16\n")
+
+except ValueError as err_version:
+    bareosfd.JobMessage(
+        bareosfd.M_FATAL,
+        f"FATAL ERROR: {err_version}\n"
+    )
+
+except ImportError as err_import:
+    bareosfd.JobMessage(
+        bareosfd.M_FATAL, f"could not import Python module: pg8000 {err_import}\n"
+    )
+
+except Exception as err_unknown:
+    bareosfd.JobMessage(
+        bareosfd.M_FATAL, f"Unknown error {err_unknown}\n"
+    )
+
+
+def parse_row(row):
+    """
+    This function exists to fix bug in pg8000 >= 1.26 < 1.30.0
+    Returned results of pg_backup_stop are string representation
+    of tuple
+    """
+    remove_parens = row[1:-1]
+    [lsn, backup_label, tablespace] = remove_parens.split(",")
+    # lsn does not have quotes
+    # remove quotes
+    backup_label = backup_label[1:-1]
+    tablespace = tablespace[1:-1]
+    return [lsn, backup_label, tablespace]
+
+
+@BareosPlugin
+class BareosFdPluginPostgreSQL(BareosFdPluginLocalFilesBaseclass):  # noqa
+    """
+    Bareos-FD-Plugin-Class for PostgreSQL online (Hot) backup of cluster
+    files and database transaction logs (WAL) archiving to allow incremental
+    backups and point-in-time (pitr) recovery.
+    If the cluster use tablespace, the backup will also backup and restore those
+    (symlinks in pg_tblspc, and real external location data)
+    The plugin will fail a job if previously done was not done on same PG major version
+    """
+
+    def __init__(self, plugindef):
+        bareosfd.DebugMessage(
+            100,
+            f"Constructor called in module {__name__} with plugindef={plugindef}\n"
+        )
+
+        bareosfd.DebugMessage(
+            100,
+            f"Python Version: {version_info.major}.{version_info.minor}.{version_info.micro}\n"
+        )
+
+        # Last argument of super constructor is a list of mandatory arguments
+        super().__init__(plugindef, ["postgresql_data_dir", "wal_archive_dir"])
+
+        self.ignore_sub_dirs = ["log", "pg_wal", "pg_log", "pg_xlog", "pgsql_tmp"]
```

Comment:
Default for 15 pg_dynshmem/, pg_notify/, pg_serial/, pg_snapshots/, pg_stat_tmp/, and pg_subtrans/